Simple Javascript Clock (Workshop Bounty)

faisalsayed10 opened the pull request:
Criteria:

- [x] Help club members produce a compelling and real project in under 20 minutes
Based on the test meeting, people were able to complete this workshop in under 20 minutes.

- [x] Are very easy to hack on for long periods of time after completion
Many instructions are listed on the end of the workshop as to how people can extend their project and learn more things.

- [x] Have instructive and compelling images, GIFs, and visuals
This workshop do contains instructive and compelling images, GIFs, and visuals.

- [x] Use repl.it or a similar service that allows members to begin coding with one click.
Repl.it is used for this workshop.

- [x] Link unknown concepts to external sources.
Linked many unknown concepts such as Javascript Dates, requestAnimationFrame etc to useful sources from where people can learn more on it.

- [x] End with at least 3 demos of how the project can be expanded, with live links and code links for each demo. 
Have provided 4 demos from the demo workshop to which people can refer and get an idea on how it can be expanded.

Good Workshop Guidelines:

- [x] Workshop that meets all of the "Good workshops" guidelines outlined above.

- [x] 1-2 minute video clip from your test club meeting showing what members built.
Here's the video clip from the demo meeting. People encountered few bugs while doing some custom coding but they got it fixed!
Link: https://hackclub.slack.com/archives/C01A9V5LPD0/p1599927330008900

- [x] All assets (images, PDFs, and any other binaries) must be hosted using the #cdn channel on the Hack Club Slack for posterity and to cut down on repository size.
All assets are hosted using the #cdn channel and gifs have been taken from giphy.

- [x] All workshops must be in English at native-level proficiency. Phrasing and grammar must be perfect.


- [x] You must be either a registered Hack Club leader or an active member of the Slack.
Club leader @ St. Mary's High School and also an active member @ slack.

zachlatta reviewed (changes requested):
Hey @faisalsayed10, this is awesome!

I just did a runthrough of the workshop myself, recording my screen as I did it. I'm currently uploading the video to https://youtu.be/YQ6javQvtK4. It should be live in about 20 minutes.

I think we're really close to merging. I started leaving some comments on GitHub with specific changes I'd like to see before merging, but then I started to just call things out in the video instead of manually commenting for each specific change.

I think we're really close. Here's what I'd like to see before merging:

1. It took me 34 minutes to finish the workshop. Can you shorten it so it only takes me 20? Keep in mind that inside of Hack Club meetings, it'll take members longer to go through the workshop than someone like me doing it alone.
2. There were a number of things, some small (like grammar) and some medium (like some explanations / areas of structure, particularly around the complex CSS and JS stuff), that I ran into in the screen recording. Can you review my screencast and do an opinionated, very detail-oriented edit of the workshop?

Once you've done both of the above, I'll be ready for a final review. This definitely qualifies for the workshop bounty program and I'd love to work with you to get it merged!

faisalsayed10 commented:
@zachlatta Here's what I did:
- [x] Watched the video and made all the necessary changes.
- [x] There was a part where a bug happened when every minute passed, fixed it.
- [x] Fixed the weird whitespaces which appeared in the code blocks. ( github's default tab_size is 8 and that is what is causing it. I've tried to fix as much as I can.
- [x] Added a short explanation + external links for some more complex concepts.

zachlatta commented:
Awesome @faisalsayed10! @Matt-Gleich is stepping up to help review workshops and manage the Hack Club GitHub repos. His first task: work with you to get this merged! I'm hoping we can get it in over the next few days.

Matt-Gleich commented:
Hey @faisalsayed10! As Zach said I am going to be helping out with workshop review. I'll check out your PR right now and hopefully, we get it merged ASAP!

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
```diff
@@ -0,0 +1,457 @@
+[final_code]: https://repl.it/@FaisalSayed1/JS-CSS-Clock
+[repl_it]: https://repl.it
+
+## Part I: Prerequisites
```
Change part numbers to normal numbers instead of Roman Numerals.

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
```diff
@@ -0,0 +1,457 @@
+
+### Setting up your code environment on Repl.it
+
+[Repl.it](https://repl.it) is an online code editor where you will build your clock today.
```
```suggestion
[Repl.it](https://repl.it) is an online code editor where you can build your clock today. You don't have to use Repl.it but we suggest you do as it sets everything up for you.
```

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
```diff
@@ -0,0 +1,457 @@
+
+### 2) The CSS file
+
+This is the `style.css` file in your directory. Whatever styles we give to our website is written in this file. If you take a look in the `<head>` tag in your html, you will find a line of code (`<link href="style.css" rel="stylesheet" type="text/css" />`). This means that your CSS file is properly linked to your HTML file!
```
```suggestion
This is the `style.css` file in your directory. Whatever styles we give to our website is written in this file. If you take a look in the `<head>` tag in your HTML, you will find a line of code (`<link href="style.css" rel="stylesheet" type="text/css" />`). This means that your HTML file is properly linked to your CSS file!
```

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
````diff
@@ -0,0 +1,457 @@
+```html
+<script src="script.js"></script>
+```
+This means that your JavaScript file is linked to your HTML.
````
```suggestion
This means that your HTML file is linked to your JavaScript.
```

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
```diff
@@ -0,0 +1,457 @@
+
+Note: Make sure the `<script>` tag, which is in your body, is below the `<div>` tags or else it will throw you errors.  
+
+This is your code so far:
```
```suggestion
This is what your code should look like so far:
```

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
````diff
@@ -0,0 +1,457 @@
+ 	const currentTime = new Date();
+}
+```
+This object contains specific methods through which we can get the current hour / minute / second.  
````
```suggestion
The `Date` object contains specific methods through which we can get the current hour / minute / second.  
```

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
```diff
@@ -0,0 +1,457 @@
+Learn more about [viewport heights](https://www.sitepoint.com/css-viewport-units-quick-start/)
+
+Next, we'll need to add some styles to our clock!  
+Following the 3 main steps given above, let's accomplish this task!  
+1. Select the class designated to the clock div.  
```
```suggestion
Following the 3 main steps given above, let's accomplish this task!
  
1. Select the class designated to the clock div.  
```

Matt-Gleich commented on workshops/Simple Js Clock/README.md:
````diff
@@ -0,0 +1,457 @@
+	secondHand.style.transform = `rotate(${secondDegrees}deg)`;
+}
+```
+**Explanation:**
````
```suggestion

**Explanation:**
```

Matt-Gleich reviewed (changes requested):
Looks great @faisalsayed10! Super approachable, fun, and informative. Just made a few very small change requests and once you implement them we should be good to merge! @zachlatta do you still want to review the latest changes?

faisalsayed10 commented:
@Matt-Gleich I have implemented the changes as you have requested.

zachlatta commented:
Can you make the folder name all lowercase and use underscores instead of spaces? Make sure you’re happy with the folder name.

From here, if this has your thumbs up @Matt-Gleich, I’d like to do one final run through tomorrow before giving mine and merging.

Congrats! And thanks for all the work you’ve done (Matt you too!).

faisalsayed10 commented:
Yes Sure!

Matt-Gleich reviewed (approved):
Looks good to me! Just fixed that description for ya but you should be good. Congrats on submitting your workshop!

Matt-Gleich commented:
@zachlatta any update on this?

zachlatta commented:
Starting my review now!

zachlatta reviewed (changes requested):
Hey @faisalsayed10, really good work here. Thank you for all the changes! I just did another workshop runthrough and have a couple final changes for you to make. Please go through it and make all requested changes.

YouTube upload is happening right now. Video at https://youtu.be/OYCciY8Depg. The ETA it's giving me is 1.5 hours until it's online.

I think we are super close. Like 95% there. Really good work on all the revisions since my last review. Once you make these final changes, I think we'll be ready to merge.

I'm going to DM you the link to the workshop payment form so you can get started on that now, since it may take a day or two. Definitely bounty worthy and looking forward to getting you paid!

faisalsayed10 commented:
Hi @zachlatta,
I watched the video and I've done all the changes you mentioned for me to do.
@Matt-Gleich Could you please help me in correcting my grammar at required places? Also, please drop in a final review by your side!

Matt-Gleich commented:
@faisalsayed10 I will review this as soon as I get out of school today at ~3PM.

faisalsayed10 commented:
Sure

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+---
+name: 'JavaScript Clock'
+description: 'Build a clock using HTML, CSS, & JavaScript.'
```
```suggestion
description: 'Build a clock using HTML, CSS, & JavaScript'
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+Let us first take a look at the HTML file (`index.html`).
+
+At line 1 we have `<!DOCTYPE html>` This declares that this file is a HTML file. If we take a look in the `<html>` tag, we will find a `<body>` tag. Here's where we will write the code. Make sure that all the code you write is included between the opening and closing body tags. (`<body></body>`).
```
```suggestion
At line 1 we have `<!DOCTYPE html>` This declares that this file is an HTML file. If we take a look in the `<html>` tag, we will find a `<body>` tag. Here's where we will write the code. Make sure that all the code you write is included between the opening and closing body tags. (`<body></body>`).
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+### 1) HTML
+
+Let us start building our clock! First we'll write the HTML code required.
```
```suggestion
Let us start building our clock! First, we'll write the HTML code required.
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+Now add 3 more `<div>` tags nested inside of the `clock-face` div. Give a class of `hand hour-hand`, `hand min-hand`, `hand second-hand` to each of the 3 new `<div>` tags respectively.
+
+**Note:** Make sure the `<script>` tag, which is in your body, is below the `<div>` tags or else it will throw you errors.
```
```suggestion
**Note:** Make sure the `<script>` tag, which is in your body, is below the `<div>` tags, or else it will throw you errors.
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+We start by giving our website a soothing green color. Go ahead and try changing it to the color of your own choice!
+
+Then we removed the default margin and padding from all the elements in our website (some elements have a default value for their margin and padding such as 2.4, 3 etc). We changed the default display `display: block` to `display: flex`.
```
```suggestion
Then we removed the default margin and padding from all the elements in our website (some elements have a default value for their margin and padding such as 2.4, 3, etc). We changed the default display `display: block` to `display: flex`.
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+A very good reason for this is because `flex` enables you to align your items perfectly to the center! And that's what we have done in the next two lines. The `align-items` property aligns all your children elements to center, but horizontally. The `justify-content` property aligns all your children elements to center, but vertically.
+
+Next we have the `min-height` property which sets a minimum height of `100vh` (viewport height) to your `<body>`. The `min-height` property works together with the `align-items` property and helps the children to align properly at the center.
```
```suggestion
Next, we have the `min-height` property which sets a minimum height of `100vh` (viewport height) to your `<body>`. The `min-height` property works together with the `align-items` property and helps the children to align properly at the center.
```

Matt-Gleich commented on workshops/simple_clock/README.md:
````diff
@@ -0,0 +1,541 @@
+```
+
+**Explanation**: 
+We gave a width and a height to our clock. Next, we gave it a border and the radius for the border. Next, is the background color for that clock (if you don't like it, consider changing it or even removing it).
````
```suggestion
We gave the width and height to our clock. Next, we gave it a border and the radius for the border. Next, is the background color for that clock (if you don't like it, consider changing it or even removing it).
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+Also, we have added some shadows to the element. We gave multiple shadows so as to get a cool effect on our clock. Also, the `inset` property changes the shadow from an outer shadow (`outset`) to an inner shadow.
+
+Now let's test our code! Click on the **RUN** button and Voila! Here's the basic background for our clock! (if it doesn't looks like this, make sure you crosscheck every single line of code).
```
```suggestion
Now let's test our code! Click on the **RUN** button and Voila! Here's the basic background for our clock! (if it doesn't look like this, make sure you crosscheck every single line of code).
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+Learn more about [transform-origin](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-origin).
+Learn more about [cubic-bezier](https://www.w3schools.com/cssref/func_cubic-bezier.asp).
+
+**Note:** The `.hand` class was given to 3 `<div>` tags which means this styling will affect all the 3 `<div>` tags. To give some extra classes to the hour hand, minute hand and the second hand specifically, we also gave those 3 `<div>` tags a unique classname.
```
```suggestion
**Note:** The `.hand` class was given to 3 `<div>` tags which means this styling will affect all the 3 `<div>` tags. To give some extra classes to the hour hand, minute hand, and the second hand specifically, we also gave those 3 `<div>` tags a unique class name.
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+## 3) JavaScript
+
+So far, we have the clock hands aligned correctly and now in this section we will animate them. Before panicking, let me ensure you that there's just 30 lines of JavaScript code!
```
```suggestion
So far, we have the clock hands aligned correctly and now in this section, we will animate them. Before panicking, let me ensure you that there are just 30 lines of JavaScript code!
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+Learn more about [query selector](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector).
+
+Now it's time to implement a function for the proper functionality of the clock. Create a function `setDate` at the bottom your file.
```
```suggestion
Now it's time to implement a function for the proper functionality of the clock. Create a function `setDate` at the bottom of your file.
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+
+If you still face difficulties in signing up watch [this](https://www.youtube.com/watch?v=Mtqp4CUepk0).
+
+Here are somethings which you can do:
```
```suggestion
Here are some things which you can do:
```

Matt-Gleich commented on workshops/simple_clock/README.md:
```diff
@@ -0,0 +1,541 @@
+Here are somethings which you can do:
+1. Consider changing the colors and adding more features!  
+2. Try to make the clock circular!  
+3. Try to make it dark theme!
```
```suggestion
3. Try to make it have a dark theme!
```

Matt-Gleich reviewed (approved):
Just fix those small little changes and we should be all good!

faisalsayed10 commented:
Fixed all the changes above and I think we should be good to merge?
cc: @zachlatta

Matt-Gleich commented:
Congrats @faisalsayed10! Your workshop has been merged.

faisalsayed10 commented:
Yay! Thanks!
